### ISSUE TITLE
[prefer-regexp-exec] False positive for union types

### ISSUE THREAD

**Issue by rajzik** [typescript-eslint/typescript-eslint#3319]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-regexp-exec": "warn",
  }
}
```

```TS
// your repro code case
const docsExlusion = ['package-lock.json', /regexp/];

const file = '';
docsExclusions.some((ex) => !!file.match(ex)); // Use the `RegExp#exec()` method instead.
```

```json
{
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "declaration": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "isolatedModules": false,
    "jsx": "preserve",
    "lib": [
      "dom",
      "DOM.Iterable",
      "esnext"
    ],
    "module": "commonjs",
    "moduleResolution": "node",
    "noEmitOnError": true,
    "noImplicitReturns": true,
    "pretty": true,
    "strict": true,
    "target": "es2018",
    "resolveJsonModule": true,
    "experimentalDecorators": true,
    "useDefineForClassFields": false,
    "allowJs": false,
    "skipLibCheck": true,
    "composite": true,
    "declarationMap": true
  }
}
```

**Expected Result**

No warning/error for this case when union type is present and autofix not applied.

**Actual Result**

Type of the variable is `string | RegExp` and rule reports warning to use RegExp method.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.22.0` |
| `@typescript-eslint/parser`        | `4.22.0` |
| `TypeScript`                       | `4.2.4` |
| `ESLint`                           | `7.24.0` |
| `node`                             | `15.11.0` |

**Comment by bradzacher** [comment]
the bug here is that the rule does not handle union types at all.
it assumes if the variable is NOT a string or a regex, then it's probably regex-like and errors.

https://github.com/typescript-eslint/typescript-eslint/blob/25ea953cc60b118bd385c71e0a9b61c286c26fcf/packages/eslint-plugin/src/rules/prefer-regexp-exec.ts#L100-L131
